Try to untangle the rid calculation.

tmds opened the pull request:
This is an attempt at simplifying how the rids get calculated.

Instead of providing a single rid, `init-distro-rid.sh` also provides a `__PortableOS` value which is the portable os that matches the target platform. It gets initialized also for non-portable builds, which enables detecting musl in the native script and passing that information along. The existing `__DistroRid` is used solely for naming the non-portable build output.

Instead of calculating the `_portableOS` of the old `__DistroRid`, `Directory.Build.props` can now directly use `__PortableOS` (and fall back to using `TargetOS`).

`_runtimeOS` is merged into `_packageOS`.

Let's see how well this works against the CI configurations. :crossed_fingers:

cc @ViktorHofer @am11 @jkotas

tmds committed:
Try to untangle the rid calculation.

tmds committed:
Ignore __portableOS.

tmds committed:
Add back musl detection.

tmds committed:
Try fix AnyOS tests.

tmds commented:
CI is looking good for these changes, with the exception of the _coreclr linux_musl_arm64_ job.

This is a cross build job, and not I'm familiar with those.

From the error message, I assume the wrong linker gets used for the crossgen2 project?
Or maybe, the build is supposed to use the crossgen2_crossarch project instead of the crossgen2 project?

@wfurt @am11 any thoughts on what changes in this PR are causing this failure?

```
  Generating native code
  /usr/bin/ld: unrecognised emulation mode: aarch64linux
  Supported emulations: elf_x86_64 elf32_x86_64 elf_i386 elf_iamcu elf_l1om elf_k1om i386pep i386pe
clang : error : linker command failed with exit code 1 (use -v to see invocation) [/__w/1/s/src/coreclr/tools/aot/crossgen2/crossgen2.csproj]
##[error]clang(0,0): error : (NETCORE_ENGINEERING_TELEMETRY=Build) linker command failed with exit code 1 (use -v to see invocation)
/__w/1/s/artifacts/transport/coreclr/build/Microsoft.NETCore.Native.targets(334,5): error MSB3073: The command ""/usr/bin/clang-15" "/__w/1/s/artifacts/obj/coreclr/crossgen2/arm64/Debug/native/crossgen2.o" -o "/__w/1/s/artifacts/bin/coreclr/linux.arm64.Debug/crossgen2/native/crossgen2" /__w/1/s/artifacts/transport/coreclr/aotsdk/libbootstrapper.a /__w/1/s/artifacts/transport/coreclr/aotsdk/libRuntime.ServerGC.a /__w/1/s/artifacts/transport/coreclr/aotsdk/libeventpipe-disabled.a /__w/1/s/artifacts/transport/coreclr/aotsdk/libstdc++compat.a /__w/1/s/artifacts/transport/coreclr/aotsdk/libnumasupportdynamic.a /__w/1/s/artifacts/bin/microsoft.netcore.app.runtime.linux-musl-arm64/Release/runtimes/linux-musl-arm64/native/libSystem.Native.a /__w/1/s/artifacts/bin/microsoft.netcore.app.runtime.linux-musl-arm64/Release/runtimes/linux-musl-arm64/native/libSystem.Globalization.Native.a /__w/1/s/artifacts/bin/microsoft.netcore.app.runtime.linux-musl-arm64/Release/runtimes/linux-musl-arm64/native/libSystem.IO.Compression.Native.a /__w/1/s/artifacts/bin/microsoft.netcore.app.runtime.linux-musl-arm64/Release/runtimes/linux-musl-arm64/native/libSystem.Net.Security.Native.a /__w/1/s/artifacts/bin/microsoft.netcore.app.runtime.linux-musl-arm64/Release/runtimes/linux-musl-arm64/native/libSystem.Security.Cryptography.Native.OpenSsl.a --sysroot=/crossrootfs/arm64 --target=aarch64-linux-gnu -g -Wl,-rpath,'$ORIGIN' -Wl,--build-id=sha1 -Wl,--as-needed -pthread -ldl -lz -lrt -lm -pie -Wl,-pie -Wl,-z,relro -Wl,-z,now -Wl,--discard-all -Wl,--gc-sections" exited with code 1. [/__w/1/s/src/coreclr/tools/aot/crossgen2/crossgen2.csproj]
##[error]artifacts/transport/coreclr/build/Microsoft.NETCore.Native.targets(334,5): error MSB3073: (NETCORE_ENGINEERING_TELEMETRY=Build) The command ""/usr/bin/clang-15" "/__w/1/s/artifacts/obj/coreclr/crossgen2/arm64/Debug/native/crossgen2.o" -o "/__w/1/s/artifacts/bin/coreclr/linux.arm64.Debug/crossgen2/native/crossgen2" /__w/1/s/artifacts/transport/coreclr/aotsdk/libbootstrapper.a /__w/1/s/artifacts/transport/coreclr/aotsdk/libRuntime.ServerGC.a /__w/1/s/artifacts/transport/coreclr/aotsdk/libeventpipe-disabled.a /__w/1/s/artifacts/transport/coreclr/aotsdk/libstdc++compat.a /__w/1/s/artifacts/transport/coreclr/aotsdk/libnumasupportdynamic.a /__w/1/s/artifacts/bin/microsoft.netcore.app.runtime.linux-musl-arm64/Release/runtimes/linux-musl-arm64/native/libSystem.Native.a /__w/1/s/artifacts/bin/microsoft.netcore.app.runtime.linux-musl-arm64/Release/runtimes/linux-musl-arm64/native/libSystem.Globalization.Native.a /__w/1/s/artifacts/bin/microsoft.netcore.app.runtime.linux-musl-arm64/Release/runtimes/linux-musl-arm64/native/libSystem.IO.Compression.Native.a /__w/1/s/artifacts/bin/microsoft.netcore.app.runtime.linux-musl-arm64/Release/runtimes/linux-musl-arm64/native/libSystem.Net.Security.Native.a /__w/1/s/artifacts/bin/microsoft.netcore.app.runtime.linux-musl-arm64/Release/runtimes/linux-musl-arm64/native/libSystem.Security.Cryptography.Native.OpenSsl.a --sysroot=/crossrootfs/arm64 --target=aarch64-linux-gnu -g -Wl,-rpath,'$ORIGIN' -Wl,--build-id=sha1 -Wl,--as-needed -pthread -ldl -lz -lrt -lm -pie -Wl,-pie -Wl,-z,relro -Wl,-z,now -Wl,--discard-all -Wl,--gc-sections" exited with code 1.

Build FAILED.
```

tmds committed:
Try fix coreclr linux_musl_arm64 job.

am11 commented on src/tests/build.proj:
```diff
@@ -435,8 +435,8 @@
       <_ConfigurationProperties>/p:TargetOS=$(TargetOS) /p:TargetArchitecture=$(TargetArchitecture) /p:Configuration=$(Configuration) /p:CrossBuild=$(CrossBuild)</_ConfigurationProperties>
-      <DotnetRestoreCommand Condition="'$(__DistroRid)' == ''">"$(DotNetTool)" restore $(RestoreProj) $(PackageVersionArg) /p:SetTFMForRestore=true $(_ConfigurationProperties)</DotnetRestoreCommand>
-      <DotnetRestoreCommand Condition="'$(__DistroRid)' != ''">"$(DotNetTool)" restore -r $(__DistroRid) $(RestoreProj) $(PackageVersionArg) /p:SetTFMForRestore=true $(_ConfigurationProperties)</DotnetRestoreCommand>
+      <DotnetRestoreCommand Condition="'$(__OutputRid)' == ''">"$(DotNetTool)" restore $(RestoreProj) $(PackageVersionArg) /p:SetTFMForRestore=true $(_ConfigurationProperties)</DotnetRestoreCommand>
+      <DotnetRestoreCommand Condition="'$(__OutputRid)' != ''">"$(DotNetTool)" restore -r $(__OutputRid) $(RestoreProj) $(PackageVersionArg) /p:SetTFMForRestore=true $(_ConfigurationProperties)</DotnetRestoreCommand>
```
These can be just `PackageRID`.
```suggestion
      <DotnetRestoreCommand Condition="'$(PackageRID)' != ''">"$(DotNetTool)" restore -r $(PackageRID) $(RestoreProj) $(PackageVersionArg) /p:SetTFMForRestore=true $(_ConfigurationProperties)</DotnetRestoreCommand>
```
(the other condition was never true since dotnet/runtime migration)

tmds replied:
Yes, I had tried it before but struggled to make the AnyOS legs work and had reverted it.
Now, I think it will work.

tmds committed:
src/tests/build.proj: use PackageRID for restore.

am11 commented:
Just a heads up: we spent days of testing in previous consolidation of RID calculation from various places in the repo. There are many scenarios not obvious and not covered by the CI. Then some folks only run their build flavors near the releases. Some edge-cases from memory:

* Visual Studio build failures (managed libraries development heavily rely on it)
* Official packages (that shows up when package validation runs / not accessible via public CI)
* Tizen; uses hybrid approach: portable build for coreclr and non-portable for libs (no source-build infra involved)
* FreeBSD / illumos: $other Unices when cross-built from linux && when built on the actual VM. Now we have good CI coverage for FreeBSD cross-built (which also implies illumos cross-build is happy). I will try to run build on their host too (once this gets closer to completion).
* Alpine Linux: cross and native build. There are old and new CI legs some of which only run during the official build.

Then we had issue with what goes into the official nuget packages and we went back and forth on fixing those.

tmds committed:
Use RuntimeOS as an override for _packageOS.

tmds committed:
Remove (probably) out-dated override for non-portable ToolsRID.

tmds committed:
CI cross-builds for linux-musl use linux.

am11 commented on Directory.Build.props:
```diff
@@ -171,82 +171,66 @@
-    <_toolsRID Condition="'$(CrossBuild)' == 'true'">$(_hostOS)-$(_hostArch)</_toolsRID>
-    <_toolsRID Condition="'$(BuildingInsideVisualStudio)' == 'true'">$(_runtimeOS)-x64</_toolsRID>
-    <_toolsRID Condition="'$(_toolsRID)' == ''">$(_runtimeOS)-$(_hostArch)</_toolsRID>
+  <!-- These are the rids used for consuming prebuilt packages. -->
```
```suggestion
  <!-- PackageRID is used for restoring packages during the build.
       e.g. during cross-build its value is different than OutputRID -->
```

am11 commented on Directory.Build.props:
```diff
@@ -171,82 +171,66 @@
-    <OutputRid Condition="'$(PortableBuild)' == 'true'">$(_portableOS)-$(TargetArchitecture)</OutputRid>
-    <TargetsLinuxBionic Condition="$(OutputRid.StartsWith('linux-bionic'))">true</TargetsLinuxBionic>
-    <TargetsLinuxMusl Condition="'$(_portableOS)' == 'linux-musl'">true</TargetsLinuxMusl>
+  <!-- OutputRid names the packages for the target platform. -->
```
```suggestion
  <!-- OutputRID is used for creating shipping packages for target platform. -->
```

am11 commented on Directory.Build.props:
```diff
@@ -171,82 +171,66 @@
+    <_outputOS>$(_parseDistroRid.SubString(0, $(_distroRidIndex)))</_outputOS>
+    <_outputOS Condition="'$(PortableBuild)' == 'true'">$(_portableOS)</_outputOS>
+
+    <OutputRid Condition="'$(OutputRid)' == ''">$(_outputOS)-$(TargetArchitecture)</OutputRid>
```
While we are at it, could you align the casing? `OutputRid` -> `OutputRID` (

tmds committed:
Add missing arch to rid.

tmds committed:
Try simplified ToolsRID.

tmds committed:
Update comments.

tmds committed:
Rename OutputRid to OutputRID.

tmds committed:
Tweak comment.

tmds commented:
> Just a heads up: we spent days of testing in previous consolidation of RID calculation from various places in the repo. There are many scenarios not obvious and not covered by the CI.

I think the rid calculation is functionally working, but it's not so clear how it is working.

If we make it easier to understand, we're also less likely to break it.

I think what's in the PR now is much clearer as to what the rids mean, and how they get calculated:

https://github.com/dotnet/runtime/blob/9c2a97759d4cc385df701fb7bc4627296dc29809/Directory.Build.props#L174-L227

There is still some weirdness in the PackageRID for CrossBuilds:

https://github.com/dotnet/runtime/blob/9c2a97759d4cc385df701fb7bc4627296dc29809/Directory.Build.props#L191-L195

That is: if you're cross-building on linux for linux-musl, PackageRID is linux-musl, and if you're cross-building on linux for freebsd, PackageRID is linux. :thinking: 

I'm keeping the status quo for this.

Besides trying to make things more clear, this PR is also fixing linux-musl detection for non-portable builds.

tmds committed:
Move TargetsLinux{Bionic,Musl} with the other Targets properties.

tmds commented:
CI is ok with these changes.

@ViktorHofer can you trigger them against a broader set of configurations?

tmds committed:
Don't use RuntimeOS for PackageRID for CrossBuilds.

tmds committed:
Move RuntimeOS condition to ToolsRID assignment.

tmds committed:
Clean up a few things.

tmds committed:
Fix broken Condition.

tmds commented:
@am11 @ViktorHofer @jkotas @akoeplinger this is up for review, ptal.

I think this is an improvement in making clear what the different rids in the calculation are, and how they get calculated.

ViktorHofer commented:
> @am11 @ViktorHofer @jkotas @akoeplinger this is up for review, ptal.

While I agree that this is an improvement, it comes with a huge risk. RID infrastructure changes tend to be viral and surface in different stages in our build (and some of those don't run per PR). Some of these [are mentioned above by @am11](https://github.com/dotnet/runtime/pull/82832#issuecomment-1452370034).

Think about validation steps that you could perform that would reduce risk, i.e. by comparing the produced package assets when building on different OSs with different runtime flavors, cross-builds, etc. Don't trust in CI validation as it only covers a partial set of impacted components.

Unfortunately I can't review or help you with these steps as I'm currently heads down on other deliverables but I will ping a few folks and maybe they or ones already pinged can take a look.

cc @ericstj @jkoritzinsky @hoyosjs

tmds commented:
> While I agree that this is an improvement, it comes with a huge risk. RID infrastructure changes tend to be viral and surface in different stages in our build (and some of those don't run per PR).

These changes keep both the input (`TargetOS`, and `RuntimeOS`) and output arguments (`ToolsRID`, `OutputRID`, and `PackageRID`) the same.

So, outputs are still placed at the same location (controlled by `TargetOS`, and `OutputRID`).
The `PackageRID` and `ToolsRID` are used for consuming packages. Getting them wrong will fail the build.

It's definitely possible these changes break something that was not tested up-front.

Because the changes untangle how the output arguments are calculated, the fixes for such breaks will be straight-forward.

I think that limits the risk. (It's the premise of this PR.)

akoeplinger commented:
This looks like a great change, I'll run an official build against it and report back.

/cc some more eyes on the RID calculation changes: @directhex @steveisok

akoeplinger commented:
The only failure on the official build is on the linux-bionic job:

```
Running 'sh -c 'build_arch="arm64" compiler="clang" . "/__w/1/s/eng/common/native/init-compiler.sh" &&  cmake -DCMAKE_INSTALL_PREFIX="/__w/1/s/artifacts/obj/mono/linux.arm64.Release/out" -DCMAKE_INSTALL_LIBDIR=lib -DCMAKE_BUILD_TYPE=Release -DGC_SUSPEND=hybrid -DMONO_LIB_NAME=monosgen-2.0 -DMONO_SHARED_LIB_NAME=coreclr -DSTATIC_COMPONENTS=1 -DMONO_COMPONENTS_RID=linux-arm64 -DCMAKE_C_FLAGS=" -Wl,--build-id=sha1" -DCMAKE_CXX_FLAGS=" -Wl,--build-id=sha1"  "/__w/1/s/src/mono"'' in '/__w/1/s/artifacts/obj/mono/linux.arm64.Release/'
  No usable version of clang found.
/__w/1/s/src/mono/mono.proj(621,5): error MSB3073: The command "sh -c 'build_arch="arm64" compiler="clang" . "/__w/1/s/eng/common/native/init-compiler.sh" &&  cmake -DCMAKE_INSTALL_PREFIX="/__w/1/s/artifacts/obj/mono/linux.arm64.Release/out" -DCMAKE_INSTALL_LIBDIR=lib -DCMAKE_BUILD_TYPE=Release -DGC_SUSPEND=hybrid -DMONO_LIB_NAME=monosgen-2.0 -DMONO_SHARED_LIB_NAME=coreclr -DSTATIC_COMPONENTS=1 -DMONO_COMPONENTS_RID=linux-arm64 -DCMAKE_C_FLAGS=" -Wl,--build-id=sha1" -DCMAKE_CXX_FLAGS=" -Wl,--build-id=sha1"  "/__w/1/s/src/mono"'" exited with code 1.
```

tmds commented:
> The only failure on the official build is on the linux-bionic job:

@akoeplinger what container image does this build use?
And what is the (top-level) `./build.sh` command in the log?

tmds committed:
Try fix the linux-bionic build.

tmds commented:
@akoeplinger instead of waiting for some more info to debug, I took a guess as what might fix the linux-bionic job. Can you give it another run?

tmds committed:
Try fix the linux-bionic build, part II.

tmds committed:
Try fix the linux-bionic build, part III.

tmds committed:
Fix PackageRID.

tmds committed:
Use RuntimeOS as _packageOS override.

tmds committed:
Remove RuntimeOS.

tmds commented:
> The only failure on the official build is on the linux-bionic job:

`linux-bionic` is special because it doesn't get detected by the native build script, so it needs to be passed in as an argument. As most things, it was handled using `RuntimeOS`.

It would be nice if we could detect it in the native build script, similar to how we're detecting `linux-musl`.
I don't know the bionic environment well enough to know if it is possible and add it.

I changed the `linux-bionic` job to pass `linux-bionic` using the `-os` argument.

I've also removed `RuntimeOS` completely, in favor of `PackageOS` and `ToolsOS`. These are only used/needed for source-build now.

akoeplinger commented:
That worked, the official build is green 🙂

akoeplinger commented on eng/common/templates/steps/source-build.yml:
```diff
@@ -63,11 +63,6 @@ steps:
-    runtimeOsArgs=
-    if [ '${{ parameters.platform.runtimeOS }}' != '' ]; then
-      runtimeOsArgs='/p:RuntimeOS=${{ parameters.platform.runtimeOS }}'
-    fi
```
we should also remove the usage of `runtimeOsArgs` on line 83, though these changes are in eng/common so they'd need to be done in the arcade repo.

akoeplinger commented on eng/SourceBuild.props:
```diff
@@ -38,7 +38,7 @@
       <InnerBuildArgs>$(InnerBuildArgs) --warnAsError false</InnerBuildArgs>
       <InnerBuildArgs>$(InnerBuildArgs) --outputrid $(TargetRid)</InnerBuildArgs>
-      <InnerBuildArgs>$(InnerBuildArgs) /p:RuntimeOS=$(RuntimeOS)</InnerBuildArgs>
+      <InnerBuildArgs>$(InnerBuildArgs) /p:PackageOS=$(RuntimeOS) /p:ToolsOS=$(RuntimeOS)</InnerBuildArgs>
```
should we rename the RuntimeOS property here too to avoid confusion?

or am I reading this right and it is overwritten by https://github.com/dotnet/installer/blob/7d35ae67bbe43ce9164074801b29e8e1bd942c27/src/SourceBuild/content/repo-projects/runtime.proj#L23

tmds replied:
Yes, source-build passes `RuntimeOS`, and we're using it to set `PackageOS` and `ToolsOS`.

akoeplinger commented:
> It would be nice if we could detect it in the native build script, similar to how we're detecting linux-musl.
> I don't know the bionic environment well enough to know if it is possible and add it.

You mean this check? https://github.com/dotnet/runtime/blob/0f1866ead5aef15d47983692bafb4ba20cd3aa98/eng/native/init-distro-rid.sh#L168-L172

linux-bionic is essentially Android without the Java parts, so we use the Android NDK to target it instead of the usual ROOTFS_DIR cross compilation. I think passing it via the `-os` value is a reasonable approach.

akoeplinger commented on Directory.Build.props:
```diff
@@ -171,89 +171,65 @@
+    <_portableOS Condition="'$(_portableOS)' == 'anyos'">$(__PortableOS)</_portableOS>
+    <_portableOS Condition="'$(_portableOS)' == 'windows'">win</_portableOS>
+    <_portableOS Condition="'$(_portableOS)' == 'linux' and '$(__PortableOS)' == 'linux-musl'">linux-musl</_portableOS>
+    <_portableOS Condition="'$(_portableOS)' == 'linux' and '$(__PortableOS)' == 'linux-bionic'">linux-bionic</_portableOS>
```
_portableOS and __PortableOS is quite easy to confuse (just a single underscore difference due to msbuild case insensitivity)

tmds replied:
I've renamed it to `__PortableTargetOS`.

akoeplinger commented on eng/build.sh:
```diff
@@ -286,6 +286,14 @@ while [[ $# > 0 ]]; do
         solaris)
           os="solaris" ;;
+        linux-bionic)
+          os="linux"
```
should this set `linux-bionic` like musl does?

tmds replied:
The `linux-musl` one is wrong. For `linux-musl` and `linux-bionic` `os` (which is `TargetOS` in the props file) is `linux`.

akoeplinger commented on Directory.Build.props:
```diff
@@ -171,89 +171,65 @@
-    <_packageOS Condition="'$(CrossBuild)' == 'true'">$(_hostOS)$(_packageLibc)</_packageOS>
-    <_packageOS Condition="'$(_packageOS)' == '' and '$(PortableBuild)' == 'true'">$(_portableOS)</_packageOS>
-    <_packageOS Condition="'$(_packageOS)' == ''">$(_runtimeOS)</_packageOS>
+  <!-- _portableOS is the portable OS corresponding to the target platform. -->
```
I think it'd be good to have an explanation comment with examples here or somewhere in the docs about what exactly the differences between PortableOS, TargetOS, PackageRID, OutputRID, ToolsRID, etc is

tmds replied:
I've added a one-line comment to each section. I'll extend it to make it more clear.

tmds replied:
I've extended the comments. ptal.

tmds committed:
Fix '-os linux-musl' mapping.

tmds committed:
Clean up .

tmds committed:
Rename __PortableOS to __PortableTargetOS.

tmds committed:
Extend comments.

tmds committed:
Extend OutputRID comment too.

tmds committed:
build.sh: include linux-bionic, linux-musl in usage.

tmds commented:
Can we do a review round on this PR, and then merge it?

tmds committed:
Merge branch 'main' into rid_cleanup

tmds commented:
@akoeplinger if there is no feedback from others, can you do another review and then merge it if it looks good?

tmds commented on eng/SourceBuild.props:
```diff
@@ -31,10 +31,12 @@
-      <!-- RuntimeOS is the build host rid OS. -->
+      <!-- PackageOS and RuntimeOS control the rids of prebuilts consumed by the build.
+           They are set to RuntimeOS so they match with the build SDK rid. -->
+      <InnerBuildArgs>$(InnerBuildArgs) /p:PackageOS=$(RuntimeOS) /p:ToolsOS=$(RuntimeOS)</InnerBuildArgs>
```
I've eliminated `RuntimeOS` in the rid calculation because it was an overloaded term with an ambiguous name.

The only user that remains is source-build, which uses `RuntimeOS` to pass the rid of the build sdk.

I kept it local to this file to make explicit source-build is the only user.
Is that ok? Or would you prefer to implement it as an override for `PackageOS` and `ToolsOS` in `Directory.Build.props`?

ViktorHofer replied:
We try to follow the [GCC configure terms](https://gcc.gnu.org/onlinedocs/gccint/Configure-Terms.html#:~:text=So%2C%20build%20is%20the%20machine,you're%20building%20libraries) which expose `Build`, `Host` and `Target` as terms to configure a build process. It would be good to follow that consistently.

Unsure if that answers your questions but I thought it might be helpful for context.

tmds replied:
I don't think we capture all 3 in `Directory.Build.props`.
If I make the analogy, I'd map _GCC Build_ to `ToolsOS`, and _GCC Target_ to `PackageOS`.
Depending on the target, appropriate things are built for the _GCC Host_.

I think we can leave `RuntimeOS` in `SourceBuild.props` (just for source-build) until we have a reason to move it to `Directory.Build.props`.

tmds commented:
I assume this is getting little feedback because the change is non-functional mostly. If someone would object, I think they'd be louder.

@akoeplinger @ViktorHofer is this good to merge?

tmds committed:
SourceBuild.props: no need to pass RuntimeOS. Only set ToolsOS, PackageOS when RuntimeOS is not empty.

tmds commented:
_ping_

tmds committed:
Fix comment.

tmds commented:
@ViktorHofer @akoeplinger can we look into getting this merged? Or if you think it is too risky, should I close the PR?

ViktorHofer commented:
Sorry for the late reply. Let's try getting this in and deal with potentially fallout.

ViktorHofer commented:
There's apparently an outage. Let's retrigger CI later today or tomorrow.

akoeplinger commented:
/azp run runtime-extra-platforms

akoeplinger commented:
The runtime-extra-platforms failures look unrelated except this one about the ilasm/ildasm packages:
```
/__w/1/s/src/libraries/Common/tests/System/TestILAssembly/TestILAssembly.ilproj : error NU1101: Unable to find package runtime.linux-bionic-x64.microsoft.netcore.ilasm. No packages exist with this id in source(s): dotnet-eng, dotnet-libraries, dotnet-public, dotnet-tools, dotnet7, dotnet7-transport, dotnet8, dotnet8-transport [/__w/1/s/Build.proj]
/__w/1/s/src/libraries/Common/tests/System/TestILAssembly/TestILAssembly.ilproj : error NU1101: Unable to find package runtime.linux-bionic-x64.microsoft.netcore.ildasm. No packages exist with this id in source(s): dotnet-eng, dotnet-libraries, dotnet-public, dotnet-tools, dotnet7, dotnet7-transport, dotnet8, dotnet8-transport [/__w/1/s/Build.proj]
/__w/1/s/src/libraries/System.Runtime/tests/TestStructs/System.TestStructs.ilproj : error NU1101: Unable to find package runtime.linux-bionic-x64.microsoft.netcore.ilasm. No packages exist with this id in source(s): dotnet-eng, dotnet-libraries, dotnet-public, dotnet-tools, dotnet7, dotnet7-transport, dotnet8, dotnet8-transport [/__w/1/s/Build.proj]
/__w/1/s/src/libraries/System.Runtime/tests/TestStructs/System.TestStructs.ilproj : error NU1101: Unable to find package runtime.linux-bionic-x64.microsoft.netcore.ildasm. No packages exist with this id in source(s): dotnet-eng, dotnet-libraries, dotnet-public, dotnet-tools, dotnet7, dotnet7-transport, dotnet8, dotnet8-transport [/__w/1/s/Build.proj]
/__w/1/s/src/libraries/System.Runtime/tests/TestModule/System.Reflection.TestModule.ilproj : error NU1101: Unable to find package runtime.linux-bionic-x64.microsoft.netcore.ilasm. No packages exist with this id in source(s): dotnet-eng, dotnet-libraries, dotnet-public, dotnet-tools, dotnet7, dotnet7-transport, dotnet8, dotnet8-transport [/__w/1/s/Build.proj]
/__w/1/s/src/libraries/System.Runtime/tests/TestModule/System.Reflection.TestModule.ilproj : error NU1101: Unable to find package runtime.linux-bionic-x64.microsoft.netcore.ildasm. No packages exist with this id in source(s): dotnet-eng, dotnet-libraries, dotnet-public, dotnet-tools, dotnet7, dotnet7-transport, dotnet8, dotnet8-transport [/__w/1/s/Build.proj]
```

tmds committed:
linux-bionic doesn't get built on linux-bionic.

tmds commented:
> The runtime-extra-platforms failures look unrelated except this one about the ilasm/ildasm packages, I don't see this one in main:

Thanks for triggering the builds and skimming through the logs.

I've pushed a change that should fix the issue. It will only affect the linux-bionic build.
Can you run it once more?

tmds committed:
Merge remote-tracking branch 'origin/main' into rid_cleanup

tmds commented:
The `runtime-linuxbionic` now failed with:
```
/__w/1/s/src/libraries/System.Runtime.InteropServices/tests/TestAssets/NativeExports/NativeExports.csproj(52,5): error MSB3073: The command "..." exited with code 1.
##[error]src/libraries/System.Runtime.InteropServices/tests/TestAssets/NativeExports/NativeExports.csproj(52,5): error MSB3073: (NETCORE_ENGINEERING_TELEMETRY=Build) The command "..." exited with code 1.
```
I'm not sure if that could be related.

I've merged with the main branch.

tmds commented:
/azp run runtime-linuxbionic

akoeplinger commented:
Yeah that failure is related, it is failing here in the `ComInterfaceGenerator.Tests.csproj`:

<img width="1402" alt="image" src="https://user-images.githubusercontent.com/1376924/233392937-7175050e-6431-4f81-bbbf-e5c788c89a3f.png">

That shouldn't happen though since we exclude the project here and it is getting excluded in main:
https://github.com/dotnet/runtime/blob/9aac64794e994021efc3d126916ff70c66c54b87/src/libraries/tests.proj#L85-L90

tmds commented:
> That shouldn't happen though since we exclude the project here and it is getting excluded in main:

I'm not sure where things go wrong, these should be true for this build: `__PortableTargetOS=linux-bionic`, `_portableOS=linux-bionic`, `TargetsLinuxBionic=true` (causing the exclude).

Can you tell from the log?
Otherwise, I'll try to reproduce the issue.

akoeplinger commented:
yeah we set that in src/libraries/Directory.Build.props, @ViktorHofer do you know why we do that?

tmds commented:
> we have SkipInferTargetOSName=true which means the TargetsLinuxBionic property assignment gets skipped.

Ah. I have moved the assignment in that group, but I didn't notice the Condition.
I'm going to move it back out.

tmds committed:
Unconditionally assign TargetsLinuxBionic/TargetsLinuxMusl.

ViktorHofer commented:
> @ViktorHofer do you know why we do that?

Yes that's intentional. We don't want to use Targets* properties under src/libraries as that conflicts with TargetPlatformIdentifier which we use these days to target specific platforms in managed code. Instead we use TargetOS.

akoeplinger commented:
/azp run runtime-linuxbionic

tmds commented:
We need to keep `TargetsLinuxBionic` and `TargetsLinuxMusl`, `TargetOS` doesn't identify them separate, they are all `Linux`.

ViktorHofer commented:
> Yeah looks like we've not followed that for e.g. TargetsMobile, TargetsAppleMobile, TargetsWasi and TargetsLinuxBionic (those are the ones I found via a quick search, there may be others). We can clean those up separately.

We had to disable these previously when we had our own `Target*s` properties under src/libraries that were based on TargetFramework. As we have been using `TargetPlatformIdentifier` and stopped using the `Targets*` properties to condition for a TargetFramework OS, for a year now, I'm fine with bringing them in again under src/libraries.

Feel free to remove this line https://github.com/dotnet/runtime/blob/614d68307c5b41d1d77a081d747c385b03ac8663/src/libraries/Directory.Build.props#L3

tmds committed:
Remove SkipInferTargetOSName.

tmds commented:
CI is looking good.

ViktorHofer commented on Directory.Build.props:
```diff
@@ -253,11 +250,13 @@
   </PropertyGroup>
 
-  <PropertyGroup Label="CalculateTargetOSName" Condition="'$(SkipInferTargetOSName)' != 'true'">
+  <PropertyGroup Label="CalculateTargetOSName">
```
Nit: Would be good to move this up in a follow-up and add TargetsMobile into the group as well.

ViktorHofer commented on src/installer/tests/Directory.Build.targets:
```diff
@@ -73,7 +73,7 @@
       <TestContextVariable Include="TEST_ARTIFACTS=$(SystemPathTestsOutputDir)" />
       <TestContextVariable Include="TEST_TARGETRID=$(TestTargetRid)" />
-      <TestContextVariable Include="BUILDRID=$(OutputRid)" />
+      <TestContextVariable Include="BUILDRID=$(OutputRID)" />
```
Nit: Should the test context variable be changed in the future to OutputRID?

ViktorHofer commented on eng/SourceBuild.props:
```diff
@@ -31,10 +31,11 @@
-      <InnerBuildArgs>$(InnerBuildArgs) /p:RuntimeOS=$(RuntimeOS)</InnerBuildArgs>
+      <!-- PackageOS and ToolsOS control the rids of prebuilts consumed by the build.
+           They are set to RuntimeOS so they match with the build SDK rid. -->
+      <InnerBuildArgs Condition="'$(RuntimeOS)' != ''">$(InnerBuildArgs) /p:PackageOS=$(RuntimeOS) /p:ToolsOS=$(RuntimeOS)</InnerBuildArgs>
```
Who provides RuntimeOS in source build?

tmds replied:
It gets set here: https://github.com/dotnet/dotnet/blob/3b38fd69120ec5cac0433923225a8d6f1d40c0bf/repo-projects/runtime.proj#L16.

ViktorHofer replied:
Is that still the right property to set or do we want to change that?

ViktorHofer reviewed (approved):
Approving with the caveat that I didn't understand all the changes in init-distro-rid.sh but if our builds are happy with the changes then so am I. Great work. Thanks a lot 💯

tmds commented:
@ViktorHofer and @akoeplinger thanks for your support!

ViktorHofer commented:
@tmat the `SourceIndexStage` leg in dotnet/runtime official builds are currently broken with the following error:

> build.cmd -subset libs.sfx+libs.oob -binarylog -os linux -ci

> D:\a\_work\1\s\.dotnet\sdk\8.0.100-preview.3.23178.7\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets(221,5): error NU5026: The file 'D:\a\_work\1\s\artifacts\bin\native\net8.0-linux-Debug-x64\libSystem.IO.Ports.Native.so.dbg' to be packed was not found on disk. [D:\a\_work\1\s\src\libraries\System.IO.Ports\pkg\runtime.linux-x64.runtime.native.System.IO.Ports.proj]
##[error].dotnet\sdk\8.0.100-preview.3.23178.7\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets(221,5): error NU5026: (NETCORE_ENGINEERING_TELEMETRY=Build) The file 'D:\a\_work\1\s\artifacts\bin\native\net8.0-linux-Debug-x64\libSystem.IO.Ports.Native.so.dbg' to be packed was not found on disk.

I will allow us 24 hours to investigate the issue, otherwise we will need to revert the change, unfortunately.

ViktorHofer commented:
I submitted a fix for it to unblock official builds: https://github.com/dotnet/runtime/pull/85215

The behavior change was unexpected. Previously, we didn't build the runtime.$(OutputRID).*.proj projects (only System.IO.Ports) in that leg as the glob pattern didn't match anything as OutputRID was set to something different:

https://github.com/dotnet/runtime/blob/46101c483f0c8d15e12d517be239c5ea6fc50af7/src/libraries/oob-src.proj#L18-L21

Now that `OutputRID` points to the `TargetOS`, the glob matches and requires the libs.native subset to be built first. Do we want to change anything is or is that the expected new behavior?

tmds commented:
> build.cmd -subset libs.sfx+libs.oob -binarylog -os linux -ci

I assume this issue is specific to this type of 'cross-build': on Windows we build for Linux.

Once we figure out what the rids were, we should be able to match the previous behavior.

> I submitted a fix for it to unblock official builds: https://github.com/dotnet/runtime/pull/85215

If you can unblock the official builds, we can fix this in the coming days.

tmds commented:
Previously, when we were building on Windows and set `TargetOS` to Linux, it would use a `_packageOS` of `win` for anything that is not `TargetsMobile`.

```xml
    <_runtimeOS Condition="'$(_runtimeOS)' == ''">$(_parseDistroRid.SubString(0, $(_distroRidIndex)))</_runtimeOS>
...
    <_runtimeOS Condition="'$(TargetsMobile)' == 'true'">$(TargetOS.ToLowerInvariant())</_runtimeOS>
...
    <_portableOS Condition="'$(_runtimeOS)' == 'win' or '$(TargetOS)' == 'windows'">win</_portableOS>
```

To go back to the previous behavior, we can set `/p:PackageOS=win` in the job, or add something like:
```xml
<_packageOS Condition="'$(_hostOS)' == 'windows' and '$(TargetsMobile)' != 'true'">win</_packageOS>
```

I assume the goal of this configuration is being able to build managed sources on Windows for non-Windows targets.

@ViktorHofer can you verify on Windows if the build works as before by adding `/p:PackageOS=win`? Should I make a PR for the `_packageOS` assignment?

ViktorHofer commented:
> I assume the goal of this configuration is being able to build managed sources on Windows for non-Windows targets.

Correct. The "source-index-stage" job which runs in our official builds is what powers https://source.dot.net. It just needs to build the managed libraries under src/libraries. We need to skip the native packages under https://github.com/dotnet/runtime/tree/main/src/libraries/System.IO.Ports/pkg as they currently require the native assets of the TargetOS to be built. Cross-building linux assets on a Windows OS isn't a supported scenario today AFAIK.

I submitted a workaround to fix the failing official builds via https://github.com/dotnet/runtime/pull/85239. My previous attempt was flawed as I missed that we can't build the libs.native subset targeting Linux, on Windows.

> @ViktorHofer can you verify on Windows if the build works as before by adding /p:PackageOS=win? Should I make a PR for the _packageOS assignment?

let me check. FWIW here's the build command: https://github.com/dotnet/runtime/blob/0219a1b8ba957a2c6492db309ab7cbd11749ef1c/eng/pipelines/runtime-official.yml#L54

tmds commented:
> Correct. The "source-index-stage" job which runs in our official builds is what powers https://source.dot.net/.

Interesting to know, I was wondering why the job was building sources like that.

> I submitted a workaround to fix the failing official builds via https://github.com/dotnet/runtime/pull/85239.

Thanks for looking into the fix.

ViktorHofer commented:
> Thanks for looking into the fix.

Unfortunately appending `/p:PackageOS=win` didn't help:
```
C:\Program Files\dotnet\sdk\8.0.100-preview.3.23178.7\Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets(
221,5): error NU5026: The file 'C:\git\runtime2\artifacts\bin\native\net8.0-linux-Debug-x64\libSystem.IO.Ports.Native.s
o.dbg' to be packed was not found on disk. [C:\git\runtime2\src\libraries\System.IO.Ports\pkg\runtime.linux-x64.runtime
.native.System.IO.Ports.proj]
```

tmds commented:
> Unfortunately, appending /p:PackageOS=win didn't help:

Yes. I didn't understand well enough what these projects were and should have taken a closer look. Somehow I had assumed they'd be generated using `PackageRID` in their name and changing it would cause it to no longer match with `OutputRID`. But they are real projects with those literal rids in their name, and to avoid the match, we should have set `OutputRID` instead.

Thanks for getting the issue fixed.

ViktorHofer commented:
> Thanks for getting the issue fixed.

Sure, but my fix was just meant as a workaround. Would it make sense to additionally add the following back?

```
<_portableOS Condition="'$(HostOS)' == 'win'">win</_portableOS>
```

MichaelSimons commented:
@tmds - This change broke the stage2 build of source-build.  The PackageRID is set to portable e.g. `linux-x64` whereas previously it was non-portable.  Looking at the changes, this looks suspect - https://github.com/dotnet/runtime/pull/82832/files#diff-9da24614831c308827a1ae533ffea392c97638c261dd42bd0f5226baa136d16eR210

```

    <!-- source-build sets PackageOS to build with non-portable rid packages that were source-built previously. -->
    <PackageRID Condition="'$(PackageOS)' != ''">$(PackageOS)-$(TargetArchitecture)</PackageRID>
    <PackageRID>$(_packageOS)-$(TargetArchitecture)</PackageRID>
```

The first PackageRID assignment will always be overridden by the second.

tmds commented:
> The first PackageRID assignment will always be overridden by the second.

Yes, this is an error.  This should have been similar to:

https://github.com/dotnet/runtime/blob/0fc8978b8f056647a88f7b18de8840fead97d16e/Directory.Build.props#L226-L228

tmds commented:
@MichaelSimons https://github.com/dotnet/runtime/pull/85350 adds the missing `Condition`.
